Add backfilled ADR (#1586)
mozilla/mdv2: docs/adr/0003-file-based-data.md

> # Use files for metrics data

Reviewer: ```suggestion
# Using files instead of a database to store aggregated metrics data
```

- Status: rejected
- Date: 2020 (this ADR was backfilled for historical purposes)

## Context and Problem Statement

The GLAM data consists mainly of a histogram, percentiles, and counts for each
set of dimensions. This can be stored in a file whose path and name can

> represent the dimensions and whose content can contain the actual data.

Reviewer: Worth mentioning the current postgres-based architecture. I recommend at least linking to your google doc:

https://docs.google.com/document/d/1JN4XQdG-8S5wbRUWAFp0XpDfqNu9ti_rdmy8OZXuEyE/edit

(that would in itself be a good ADR, though I understand if you don't have time to reformat it)

## Decision Drivers

- The data should be efficient to look up and read for the front-end
- The data should be efficient to write for the ETL
- The data should be cost efficient to write and maintain

## Considered Options

> - Move aggregation data, the output from the ETL, into a Postgresql database

Reviewer: ```suggestion
- Keep aggregated data, the output from the ETL, in a Postgresql database
```

- Move aggregation data into another file format more usable for the GLAM
  front-end

## Decision Outcome

> Moving the data into Postgresql was chosen, as outlined in the pros/cons below.

Reviewer: ```suggestion
Keeping the data in Postgresql was chosen, as outlined in the pros/cons below.
```

# cat glam-file-XXX.csv.gz |gunzip | sqlite3 -csv -separator ',' glam.db '.import /dev/stdin glam'
# This script runs *much* faster if you create an index on the table:
# create index metrics_index on glam(metric, channel, process, os)

import json
import os
import sqlite3

sqliteConnection = sqlite3.connect('glam.db')
cursor1 = sqliteConnection.cursor()
cursor2 = sqliteConnection.cursor()

for (metric, channel, process, operating_system) in cursor1.execute("SELECT distinct metric, channel, process, os from glam"):
    outdir = os.path.join('data', metric, channel, process, operating_system)
    os.makedirs(outdir, exist_ok=True)
    aggregates = cursor2.execute(
      "SELECT app_build_id, aggregates from glam where metric=? AND channel=? AND process=? AND os=?",
      (metric, channel, process, operating_system)
    ).fetchall()
    open(os.path.join(outdir, 'data.json'), 'w').write(json.dumps(aggregates))
```

#### Pros/cons:

- Pro: This could be a simpler local dev set up.
- Pro: This could improve the slow import times due to the materialized view.
- Pro: This could make adding new products to glam a bit simpler -- no need to

>   create new Django models.

Reviewer: Is this actually true? I don't see why we would *need* seperate models (tables) per application with the current approach (although maybe it makes more sense?).

- Con: Since data outside the range of the ETL needs to be kept, this can't be a
  simple write operation -- the fresh data needs to be added or updated while
  keeping older data and the resulting data needs to be written out. This
  complicates the file based architecture.

### Database aggregate storage

> With this approach we would write the data to an unindexed Postgresql table to

Reviewer: This is written as if it's an option we're considering, rather than the current state of things, which is a bit confusing.

```suggestion
This is the current approach. We write the data to an unindexed Postgresql table to
```
